lib/wcgi-runner/src/context.rs

```diff
@@ -0,0 +1,226 @@
+            .capabilities(Capabilities {
+                insecure_allow_all: true,
+                http_client: HttpClientCapabilityV1::new_allow_all(),
+            })
+            .sandbox_fs(self.fs()?)
+            .preopen_dir(Path::new("/"))?;
+
+        let done = self
+            .tokio_handle
+            .spawn_blocking(move || builder.run(module))
+            .map_err(Error::from)
+            .and_then(|r| async { r.map_err(Error::from) });
+
+        let handle = self.tokio_handle.clone();
+        self.tokio_handle.spawn(async move {
+            if let Err(e) =
+                drive_request_to_completion(&handle, done, body, req_body_sender, stderr_receiver)
+                    .await
+            {
+                tracing::error!(
+                    error = &e as &dyn std::error::Error,
+                    "Unable to drive the request to completion"
+                );
+            }
+        });
+
+        let mut res_body_receiver = tokio::io::BufReader::new(res_body_receiver);
+
+        let parts = dialect
+            .extract_response_header(&mut res_body_receiver)
+            .await?;
+        let chunks = futures::stream::try_unfold(res_body_receiver, |mut r| async move {
+            match r.fill_buf().await {
+                Ok(chunk) if chunk.is_empty() => Ok(None),
+                Ok(chunk) => {
+                    let chunk = chunk.to_vec();
+                    r.consume(chunk.len());
+                    Ok(Some((chunk, r)))
+                }
+                Err(e) => Err(e),
+            }
+        });
+        let body = hyper::Body::wrap_stream(chunks);
+
+        let response = hyper::Response::from_parts(parts, body);
+
+        Ok(response)
+    }
+
+    fn fs(&self) -> Result<TmpFileSystem, Error> {
+        let root_fs = RootFileSystemBuilder::new().build();
+
+        if !self.mapped_dirs.is_empty() {
+            let fs_backing: Arc<dyn FileSystem + Send + Sync> =
+                Arc::new(PassthruFileSystem::new(wasmer_wasi::default_fs_backing()));
+
+            for (src, dst) in &self.mapped_dirs {
+                let src = match src.starts_with('/') {
+                    true => PathBuf::from(src),
+                    false => Path::new("/").join(src),
+                };
+                tracing::trace!(
+                    source=%src.display(),
+                    alias=%dst.display(),
+                    "mounting directory to instance fs",
+                );
+
+                root_fs
+                    .mount(PathBuf::from(&src), &fs_backing, dst.clone())
+                    .map_err(|error| Error::Mount {
+                        error,
+                        src,
+                        dst: dst.to_path_buf(),
+                    })?;
+            }
+        }
+        Ok(root_fs)
+    }
+}
+
+/// Drive the request to completion by streaming the request body to the
+/// instance and waiting for it to exit.
+async fn drive_request_to_completion(
```

Comment:
Don't we have tooling for this in `wcgi-host`?

Reply:
I guess we could pull it into the `wcgi-host` crate, but I feel like the code is going to be closely coupled enough to the way our `wasmer_wasi` implementation works that it won't be a very clean abstraction. 

In particular, there is enough juggling going on with the various pipes, setting up the `WasiEnvBuilder`, waiting on exit codes, and such that there's no easy way to abstract out the process of communicating with a WCGI binary. I feel like we'd end up creating a trait that only ever has one implementation because it's just not expressive enough to encompass everyone's needs.